File: shared/ui/ipc/host.protocol.review.ts
import { RequestType } from "vscode-jsonrpc";
import { IpcRoutes } from "./webview.protocol";

export type ReviewCheckpoint = number | undefined;

export interface ReviewShowDiffRequest {
	reviewId: string;
	checkpoint: ReviewCheckpoint;
	repoId: string;
	path: string;
}

export interface ReviewShowDiffResponse {}

export const ReviewShowDiffRequestType = new RequestType<
	ReviewShowDiffRequest,
	ReviewShowDiffResponse,
	void,
	void
>(`${IpcRoutes.Host}/review/showDiff`);

export interface ReviewShowLocalDiffRequest {
	repoId: string;
	path: string;
	includeSaved: boolean;
	includeStaged: boolean;
	baseSha: string;
}

export interface ReviewShowLocalDiffResponse {}

export const ReviewShowLocalDiffRequestType = new RequestType<
	ReviewShowLocalDiffRequest,
	ReviewShowLocalDiffResponse,
	void,
	void
>(`${IpcRoutes.Host}/review/showLocalDiff`);

export interface ReviewCloseDiffRequest {}

export interface ReviewCloseDiffResponse {}

// this doesn't work yet due to limitations in vscode window handling
export const ReviewCloseDiffRequestType = new RequestType<
	ReviewCloseDiffRequest,
	ReviewCloseDiffResponse,
	void,
	void
>(`${IpcRoutes.Host}/review/closeDiff`);

export interface TraverseDiffsRequest {
	direction: "next" | "previous";
}

export interface TraverseDiffsResponse {}

export const TraverseDiffsRequestType = new RequestType<
	TraverseDiffsRequest,
	TraverseDiffsResponse,
	void,
	void
>(`${IpcRoutes.Host}/review/diffs/traverse`);

export interface ShowPreviousChangedFileRequest {}
export interface ShowNextChangedFileRequest {}
export interface ShowChangedFileResponse {}

export const ShowPreviousChangedFileRequestType = new RequestType<
	ShowPreviousChangedFileRequest,
	ShowChangedFileResponse,
	void,
	void
>(`${IpcRoutes.Host}/review/changedFiles/previous`);

export const ShowNextChangedFileRequestType = new RequestType<
	ShowNextChangedFileRequest,
	ShowChangedFileResponse,
	void,
	void
>(`${IpcRoutes.Host}/review/changedFiles/next`);

File: vs/src/CodeStream.VisualStudio.Core/Models/Host.Protocol.Review.cs
namespace CodeStream.VisualStudio.Core.Models {
	public class ReviewShowLocalDiffRequest {
		public string RepoId { get; set; }
		public string Path { get; set; }
		public bool? IncludeSaved { get; set; }
		public bool? IncludeStaged { get; set; }
		public string EditingReviewId { get; set; }
		public string BaseSha { get; set; }
	}

	public class ReviewShowLocalDiffRequestType : RequestType<ReviewShowLocalDiffRequest> {
		public const string MethodName = "host/review/showLocalDiff";
		public override string Method => MethodName;
	}

	public class ReviewShowDiffRequest {
		public string ReviewId { get; set; }
		public int? Checkpoint { get; set; }
		public string RepoId { get; set; }
		public string Path { get; set; }
	}

	public class ReviewShowDiffRequestType : RequestType<ReviewShowDiffRequest> {
		public const string MethodName = "host/review/showDiff";
		public override string Method => MethodName;
	}

	public class ReviewCloseDiffRequest { }
	public class ReviewCloseDiffRequestType : RequestType<ReviewCloseDiffRequest> {
		public const string MethodName = "host/review/closeDiff";
		public override string Method => MethodName;
	}

	public class ShowPreviousChangedFileRequest { }

	public class ShowPreviousChangedFileRequestType : RequestType<ShowPreviousChangedFileRequest> {
		public const string MethodName = "host/review/changedFiles/previous";
		public override string Method => MethodName;
	}

	public class ShowNextChangedFileRequest { }

	public class ShowNextChangedFileRequestType : RequestType<ShowNextChangedFileRequest> {
		public const string MethodName = "host/review/changedFiles/next";
		public override string Method => MethodName;
	}
}
